### ISSUE TITLE
Add WebAssembly lib

### ISSUE THREAD

**Issue by kitsonk** [microsoft/TypeScript#29747]
## Search Terms

wasm webassembly types lib

## Suggestion

In #20062 (locked due to age) it was discussed that WebAssembly was "experimental" and shouldn't be added as a standard lib.  That of course was in Nov 2017.  It [really isn't the case](https://developer.mozilla.org/en-US/docs/WebAssembly#Browser_compatibility) anymore.

## Use Cases

Using Definitely Typed and `@types` has certain challenges and limitations.  Where as a `lib` distributed with the typescript, based on the [IDL](https://webassembly.github.io/spec/js-api/#idl-index) would ensure it was easier to consume those APIs, similar to the web worker lib. 

## Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by trusktr** [comment]
There's package for it, https://www.npmjs.com/package/@types/webassembly-js-api, but it is marked as deprecated.

**Comment by Pauan** [comment]
The `@types/webassembly-js-api` package doesn't even work anymore. It fails to install with the following error:

```
error Couldn't find package "webassembly-js-api@*" required by "@types/webassembly-js-api" on the "npm" registry.
```

So that means it's not currently possible to use WebAssembly with TypeScript (unless you want to write your own declaration files).

I am a part of the Rust WebAssembly Working Group, our goal is to have seamless integration between Rust and the JavaScript ecosystem.

Right now, wasm-bindgen can take Rust code and compile it into `.wasm` + `.js` + `.d.ts`, which allows for seamlessly using Rust code within a JS or TS project.

Except it can't be used in a TS project because of the lack of typings for `WebAssembly`. We would really like to have seamless support for TS, but this is blocking us.

**Comment by jhenninger** [comment]
TypeScript 3.5.1 includes [type definitions for WebAssembly](https://github.com/microsoft/TypeScript/blob/v3.5.1/lib/lib.dom.d.ts#L17677). That's why `@types/webassembly-js-api` was deprecated. 

This issue can probably be closed.

**Comment by Pauan** [comment]
@jhenninger That's wonderful news! It's unfortunate that those using older versions of TypeScript (which is 99+% of TS users) have no solution, so it would have been nice if `@types/webassembly-js-api` could continue to work for a while.

**Comment by jhenninger** [comment]
@Pauan Try installing the version before it was deprecated: `npm i @types/webassembly-js-api@0.0.3`

**Comment by kitsonk** [comment]
I wouldn't have put them in `dom` personally, since it isn't specifically a `dom` API, but hey ho, at least they are there.

Fixed in f957f39cf3233752430842d6aa83aa0288d971ab
